File: ga/mutator.go
package main

import (
    "math/rand"
)

type Mutator interface {
    Mutate(individual Individual) []byte
}

type RandomReset struct {
    mutationRate uint
}

func (mutator RandomReset) Mutate(individual Individual) []byte {
    randomNumber := rand.Intn(100)
    if uint(randomNumber) > mutator.mutationRate {
        return individual.genome
    }
    randomByte := randStringRunes(1)
    randomNumber = rand.Intn(len(individual.genome))
    individual.genome[randomNumber] = randomByte[0]
    return individual.genome
}

File: ui/src/lib/ts/ga/mutator.ts
import type { Individual } from "./individual";
import { Utils } from "./utils";

export interface Mutator {
  mutate(individual: Individual): string[];
}

export class RandomReset implements Mutator {
  mutationRate: number;

  constructor(mutationRate: number) {
    this.mutationRate = mutationRate;
  }

  mutate(individual: Individual): string[] {
    [...individual.genome].forEach((char, i) => {
      if (Math.random() > this.mutationRate) {
        return;
      }
      individual.genome[i] = Utils.generateRandomChars(1)[0];
    });
    return individual.genome;
  }
}

export class CustomMutator implements Mutator {
  mutate(individual: Individual): string[] {
    throw new Error("Custom mutate not implemented");
  }
}